### PR TITLE
Return SAST scans for all projects w/in component

### PR DESCRIPTION
**Pull request by tatlax3636**
New endpoint to get most recent scan for each project within a component

### CODE REVIEW

File: src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java [hunk 1]
```diff
@@ -68,6 +69,33 @@ public CodeQualityServiceImpl(CodeQualityRepository codeQualityRepository,
...
+            CodeQuality tempCQ = codeQualityRepository.findTop1ByCollectorItemIdOrderByTimestampDesc(item.getId());
+            String name = (String) item.getOptions().get("projectName");
+            String reportUrl = (String) item.getOptions().get("reportUrl");
+
+            // If name is null, use the first part of the description
+            if(name.equals("") || name.equals(null)) {
+                name = item.getDescription().split(":")[0];
+            }
+
+            tempCQ.setName(name);
```
Comment (nameisaravind) [inline src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java:91]: @tatlax3636 plz add null check for tempCQ.

File: src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java [hunk 2]
```diff
@@ -68,6 +69,33 @@ public CodeQualityServiceImpl(CodeQualityRepository codeQualityRepository,
...
+
+        ArrayList<CodeQuality> codeQualities = new ArrayList<CodeQuality>();
+
+        collectorItems.forEach(item -> {
+            CodeQuality tempCQ = codeQualityRepository.findTop1ByCollectorItemIdOrderByTimestampDesc(item.getId());
+            String name = (String) item.getOptions().get("projectName");
+            String reportUrl = (String) item.getOptions().get("reportUrl");
+
+            // If name is null, use the first part of the description
+            if(name.equals("") || name.equals(null)) {
```
Comment (nameisaravind) [inline src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java:87]: @tatlax3636 possible NPE. reverse the equals check values.

File: src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java [hunk 3]
```diff
@@ -68,6 +69,32 @@ public CodeQualityServiceImpl(CodeQualityRepository codeQualityRepository,
...
+        ArrayList<CodeQuality> codeQualities = new ArrayList<CodeQuality>();
+
+        collectorItems.forEach(item -> {
+            CodeQuality tempCQ = codeQualityRepository.findTop1ByCollectorItemIdOrderByTimestampDesc(item.getId());
+            String name = (String) item.getOptions().get("projectName");
+            String reportUrl = (String) item.getOptions().get("reportUrl");
+
+            // If name is null, use the first part of the description
+            if(name == "" || name == null) {
+                name = item.getDescription().split(":")[0];
```
[the remark this answers is not in the document]
Comment (tatlax3636) [reply src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java:88]: @muse-dev help

[the remark this answers is not in the document]
Comment (TomMD) [reply src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java:88]: I imagine you actually wanted help on the message, we're going to be making the bug title (StringSplitter here) a link to further documentation.  In this case the report comes from error prone and has more information [here](https://errorprone.info/bugpattern/StringSplitter).

File: src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java (hunk 1 above)
Comment (tatlax3636) [reply src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java]: Added

File: src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java (hunk 2 above)
Comment (tatlax3636) [reply src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java]: Updated

File: src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java [hunk 4]
```diff
@@ -68,6 +71,38 @@ public CodeQualityServiceImpl(CodeQualityRepository codeQualityRepository,
...
+
+        collectorItems.forEach(item -> {
+            CodeQuality tempCQ = codeQualityRepository.findTop1ByCollectorItemIdOrderByTimestampDesc(item.getId());
+
+            if(tempCQ != null){
+                String name = (String) item.getOptions().get("projectName");
+                String reportUrl = (String) item.getOptions().get("reportUrl");
+
+                // If name is null, use the first part of the description
+                if( name.equals("") ) {
```
Comment (rvema) [inline src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java:91]: can cause npe please use StringUtils.isEmpty()

File: src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java [hunk 5]
```diff
@@ -68,6 +71,38 @@ public CodeQualityServiceImpl(CodeQualityRepository codeQualityRepository,
...
+        Component component = componentRepository.findOne(request.getComponentId());
+
+        Iterable<CollectorItem> collectorItems = component.getCollectorItems(request.getType().collectorType());
+
+        ArrayList<CodeQuality> codeQualities = new ArrayList<CodeQuality>();
+
+        collectorItems.forEach(item -> {
+            CodeQuality tempCQ = codeQualityRepository.findTop1ByCollectorItemIdOrderByTimestampDesc(item.getId());
+
+            if(tempCQ != null){
```
Comment (rvema) [inline src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java:86]: could be restructured, lets avoid huge if blocks

File: src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java [hunk 6]
```diff
@@ -68,6 +71,38 @@ public CodeQualityServiceImpl(CodeQualityRepository codeQualityRepository,
...
+    @Override
+    public Iterable<CodeQuality> getAllSecurityScans(CodeQualityRequest request) {
+
+        Component component = componentRepository.findOne(request.getComponentId());
+
+        Iterable<CollectorItem> collectorItems = component.getCollectorItems(request.getType().collectorType());
+
+        ArrayList<CodeQuality> codeQualities = new ArrayList<CodeQuality>();
+
+        collectorItems.forEach(item -> {
```
Comment (rvema) [inline src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java:83]: can this transformation happen on the client side?

**Review by rvema — CHANGES_REQUESTED** [review]
please address the review comments

File: src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java (hunk 4 above)
Comment (tatlax3636) [reply src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java]: Fixed

File: src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java (hunk 5 above)
Comment (tatlax3636) [reply src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java]: Restructured

File: src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java (hunk 6 above)
Comment (tatlax3636) [reply src/main/java/com/capitalone/dashboard/service/CodeQualityServiceImpl.java:83]: Discussed with Ragha, decided to leave it here and change the endpoint and function names to indicate it's only for UI use

**Review by rvema — APPROVED** [review]
lgtm

**Pull request merged**